File: step1/leetcode/cpp/tree_lc100_tree_isSameTree.cpp
#include <iostream>
#include <string>
using namespace std;
// Definition for a binary tree node.
 struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if (p == NULL && q == NULL){
            return true;
        }
        else if(p == NULL || q == NULL){
            return false;
        }
        else if(p->val != q->val){
            return false;
        }
        else{
            return isSameTree(p->left,q->left) && isSameTree(p->right,q->right);
        }
    }
};

int main(){
    TreeNode* p = new TreeNode(1);
    p->left = new TreeNode(NULL);
    p->right = new TreeNode(2);

    TreeNode* q = new TreeNode(1);
    q->left = new TreeNode(NULL);
    q->right = new TreeNode(2);

    Solution* sln = new Solution();
    bool ans = sln->isSameTree(p,q);
    cout << ans << endl;
    return 0;
}

File: step1/leetcode/python/tree_lc100_tree_isSameTree.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
    
def PreorderTraversal(p,elem_list):
    """
    :type p: TreeNode
    """
    if p :
        print(p.val)
        PreorderTraversal(p.left,elem_list)
        PreorderTraversal(p.right,elem_list)

class Solution(object):
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """

        if p == None and q == None:
            return True
        elif p == None or q == None: 
            return False
        elif p.val != q.val:
            return False
        else:
            return  self.isSameTree(p.left,q.left) and self.isSameTree(p.right,q.right)


if __name__ == "__main__":
    sln = Solution()

    p = TreeNode(1)
    p.left = TreeNode(2)

    q = TreeNode(1)
    q.left = TreeNode(None)
    q.right = TreeNode(2)

    ans = sln.isSameTree(p,q)
    print(ans)

        
        